create Division ID Governance OCDEP

djbridges opened the pull request:
A proposed governance model for OCD division IDs, as discussed in the Ballot Data Working Group

djbridges committed:
Create ocdid_governance

djbridges committed:
Rename ocdid_governance to ocdid_governance.rst

djbridges committed:
style update test

djbridges committed:
Update ocdid_governance.rst

djbridges committed:
format general contribution

djbridges committed:
finish stylizing

djbridges committed:
date change

djbridges committed:
bullet point fixes

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+``User``
+~~~~~~~~
+Anyone using OCDIDs may contribute through Slack/Google group discussion, reporting issues, or assisting with subject-matter expertise.
+These channels for communication will be publicaly displayed on the Open Civic Data website.
```
s/publicaly/publicly/

What do you mean by "open civic data" website? opencivicdata.org (which is controlled by @jamesturk) currently redirects to http://docs.opencivicdata.org/en/latest/

Do you mean docs.opencividata.org or something else?

djbridges replied:
Yes, I meant having them in the docs. We had also discussed creating a more friendly, splash-pagey introduction to OCDIDs for folks who are just discovering them (especially government folks) where they could be listed, but I see the docs being the place that gets canonically updated.

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+------------------------
+``User``
+~~~~~~~~
+Anyone using OCDIDs may contribute through Slack/Google group discussion, reporting issues, or assisting with subject-matter expertise.
```
Let's not name the particular technologies (i.e. Slack, Google Group) that we are using now, as that might change in five years.

djbridges replied:
sgtm

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+``Contributor``
+~~~~~~~~~~~~~~~
+Any individual or organization ``user`` may become a ``contributor`` by submitting a pull request adding, correcting, or aliasing jurisdictional OCDIDs.
+First time ``contributors`` will be advised to consult with an existing ``committer`` before submitting their first PR.
```
How will the be advised? Throught a CONTRIBUTING.RST file or some other mechanism?

djbridges replied:
yep

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+~~~~~~~~~~~~~
+``Committers`` have the ability to approve and commit pull requests from ``contributors`` and other ``committers``. 
+An initial cohort of ``committers`` will be determined by current project contributors.
+Any individual or organization who is a ``contributor`` and who agrees to the repsonsibilities listed here may apply to become a ``committer``. 
```
s/repsonsibilties/responsibilities/

What does it mean to "apply"?

fgregg replied:
Like send an e-mail to the mailing list?

djbridges replied:
"Apply" is maybe too heavy a word here. Going with "request" instead. And, yeah, I'm imagining they send an email/post in a slack channel/whatever we're doing in five years, then that request will get forwarded to the group of current contributors to see if there are any objections. Making that more explicit.

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+Any individual or organization who is a ``contributor`` and who agrees to the repsonsibilities listed here may apply to become a ``committer``. 
+
+* There will be a two month period where existing ``committers`` may raise objections to an applicant, otherwise the applying organization/individual will be considered approved.
+* In the case that an objection is raised, a discussion and vote will be had amongst existing ``committers``, with a simple majority necessary for approval.
```
Need some definition of a quorum probably.

djbridges replied:
Sure

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+
+``Committers`` will be approved on an organizational basis, unless explicitly noted for an individual.
+
+* Multiple members of an organization may have committing ability, but the organization itself will be counted for all voting and approval processes.
```
Will every individual with a commit bit need to be approved be current committers?

jpmckinney replied:
GitHub operates on an individual level. It's unmanageable (and probably undesirable) to add an entire organization as committers (e.g. Google). If we are operating on an organizational basis, the organization should coordinate with committers which (small number of) people will be committers, and should notify committers when the individual leaves the organization… What's the (real) problems with individuals only? Bodies like the W3C have individuals on working groups representing organizations, but they are still individuals. The working groups don't have organizational membership.

jamesturk replied:
+1 to that, let's avoid 'organizations' - the one issue I see is that the delay in approval might hold an org up, but that could either be addressed through asking for an expedited decision or encouraging each org to have two committers

djbridges replied:
My concerns were mostly around giving organizations the flexibility to designate/manage their own committers for efficiency's sake, and also ensuring that an organization with a large number of members couldn't wield outsized voting power. I'm fine going individual only and letting community norms/processes be a check.

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+* Multiple members of an organization may have committing ability, but the organization itself will be counted for all voting and approval processes.
+* Organizations will be asked to designate a primary point of contact who will act on behalf of the organization in procedural matters.
+
+``Committers`` are expected to participate in approval, support, maintenance, and other community related activities or may have their ``committer`` status revoked. 
```
How would revocation happen?

djbridges replied:
Adding in a detailed process mostly borrowed from http://docs.openvswitch.org/en/latest/internals/committer-grant-revocation/

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+
+``Committers`` are expected to participate in approval, support, maintenance, and other community related activities or may have their ``committer`` status revoked. 
+
+* ``Committers`` will have their status reviewed on a calendar year basis. 
```
By whom. All committers will be reviewed by all other committers?

jamesturk replied:
agree that this either needs to be fleshed out or omitted

djbridges replied:
heard

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+* If the contribution is from a current ``committer`` in good standing, only one additional ``committer`` review is necessary.
+* No two members of the same organization may be involved in the acceptance of a pull request.
+
+Commits should be reviewed within 2 weeks of a pull request. Accellerated timeline needs should be communicated via Slack/email.
```
s/Accellerated/Accelerated/

Again, let's not name specific technology (Slack)

djbridges replied:
ack

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+
+Commits should be reviewed within 2 weeks of a pull request. Accellerated timeline needs should be communicated via Slack/email.
+
+* If, after two months there is neither a formal approval or an ongoing conversation around a request, that request will be considered accepted.
```
Even if the person is not a committer, but just a contributor?

jpmckinney replied:
This seems dangerous. I understand that timely review is a present issue, but once we implement the rest of the steps, it should be largely mitigated. (Introducing this raises the worst-case scenario of being bound to accept a PR that deletes all files.)

djbridges replied:
@fgregg I'm thinking yes, as long as they're not a new contributor (which shouldn't be an issue, assuming the new committer is following the 'consult with a current committer' guideline in the COMMITTING.rst). I'm imagining the scenario where this most often would go into effect for a contributor would be for minor changes/corrections, as folks who would be making larger changes would probably either want to become or will be part of an organization that at least has a committer. If you think that's a bad assumption, or have strong feelings otherwise, happy to continue discussing.

djbridges replied:
@jpmckinney I hear you and agree that it should be largely mitigated, but still think having an explicit backstop against things languishing in PR-purgatory is important. What if instead of straight auto-approve, a PR that hasn't been touched in 2 months gets pro-actively sent to the whole committer group for a "speak now or forever hold your peace" style review? It can still be approved by silence, if, say, it's about an esoteric thing that no current committers have knowledge of/an opinion on, but it ensures there are eyes on it to avoid a "delete everything" scenario?

jpmckinney replied:
I'm not familiar with explicit backstops in open source development. Is there precedent for that? (preferably larger projects)

If literally no committer has an interest or opinion, I think that's to be solved by adding more knowledgeable committers – not by silently approving PRs. Divisions are political. Let's take fictional Kharun and South Kharun from the show Borgen. We get a PR that puts a division in the hierarchy of one country. But in fact, it's contested territory. I'd rather have someone with  political acuity chime in, than approve that change just because no present committer cares.

In the real world, if we got a PR for `ocd-division/country:cn/province:tw` for example (making Taiwan a province of China), it would at least make me think before merging. Whether we'd issue both that and `ocd-division/country:tw` to satisfy both views on the matter, I don't know, but I know it'd require some care, and shouldn't be automatic.

djbridges replied:
Those are both interesting cases that I agree would be tricky to handle, though even in the construction I mentioned above could still be prevented from getting automatically accepted by someone saying "I'm not sure what's going on here but maybe we should have a longer conversation/reach out to an expert." 

What I see as a more common way this would arise, though admittedly it may only happen in the US, is this: a researcher's gets on the phone with Elko, Nevada and discovers they have a separately elected [television control board](http://www.elkocountynv.net/departments/clerk/2012%20TV%20District%208%20of%209.pdf) that not only has it's own distinct geography, but it has [geographic subdistricts](https://www.dropbox.com/s/ow4c4u601a9n4gd/Screenshot%202017-11-01%2019.15.04.JPG?dl=0) from which people are elected. They create``ocd-division/country:us/state:nv/county:elko/television:e`` (or some such) and want to add it to the repo. Repeat any number of times for any number of weird districts, precinct name changes, etc. I completely get why someone wouldn't feel comfortable giving that an affirmative approval on face, but at some point it starts affecting how comprehensive tools using OCDIDs can be.

If there's a better way of handling that sort of use case that I'm missing, I'm all for it. Otherwise I think the sheer volume of those cases, in the US at least,  means the benefit of allowing them a graceful way to get in outweighs the potential for harm.

jpmckinney replied:
I think this is partly solved by having (geographically scoped) committers who are more lax with respect to performing verification and who are happy to merge things from people who appear to have good knowledge of a subject area. Basically, I still think this problem is solved at the level of people (who is deciding to merge), not at the level of process (when to auto-merge, etc.).

As we progress further down the hierarchy, the consequences of a bad merge become more and more limited. So, I'm fine with the governance of television control boards divisions being, "You say it exists? Sure! I'll merge."

We have some identifiers for countries for which no committer has sophisticated knowledge, and we have no identifiers for most countries. I think, for all such countries, we need an exemption from any sort of process-based auto-merge-like rule, and we also need an exemption for the top-level of the OCDID hierarchy (`ocd-division/country:`). Failure to adequately protect those with people-based governance can severely damage the project.

I'd also be in favor of scoping the auto-merge-like rule to specific geographies and/or specific division types (like `television`). I don't think the type `state` in the US should fall under the auto-merge-like rule – but I'll let US committers decide. For Canada, there's very low demand for PRs, so I'm not sure I need to implement such a rule at this time.

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+
+Approval will be handled on a per-file, rather than a per-commit, basis.
+
+New commits should include have a well-formed explanations, especially if generating new IDs or types.
```
This sentence got a little mangled.

djbridges replied:
Proofreading is fine for some. I like to live dangerously.

fgregg commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+
+New commits should include have a well-formed explanations, especially if generating new IDs or types.
+
+Formalized guidelines for approval will be created for use by ``committers``. These guidelines will include:
```
We'll these be automated checks.

djbridges replied:
I would love for as many as possible to be, and think that there is a lot of opportunity for automated validation. I think there's always going to be a little bit of need for a human touch though, especially for things like determining whether a new flavor of special district is distinct enough to deserve its own type and the like. I also think there should still be more rigorous guidelines written before these automated checks get implemented.

fgregg reviewed (changes requested):
Hi Donny,

Thanks for getting this started. Looks like a great start. Here's my initial feedback.

jpmckinney commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,113 @@
+--------
+``Committers`` will be required to participate in a quarterly review of new OCDIDs in order to ensure quality on-going.
+``Committers`` will be requested to contribute and maintain an ongoing style guide for creating new district types.
+``Committers`` will be required to participate in > 60% of all formal votes/actions as announced.
```
I think we need there to be geographic segmentation of committers. A committer for Antarctica shouldn't have to participate in US votes/actions.

jamesturk replied:
agree.  I think updating this document w/ committers and their jurisdiction would be useful, perhaps ~3 tiers

- a few people might be considered 'global' and their participation could be requested in all votes (but we'd presumably at least informally waive the voting requirement?)
- most people would be scoped to a country
- any gov't officials would probably be scoped to their jurisdiction and thus not expected to participate in wider conversations

djbridges replied:
Completely agree on the need for geographic segmentation, and like the tiers. Also adding in a caveat about flexibility of process/governance at a country level.

jpmckinney reviewed (changes requested):
Generally good! Please spell-check (many typos besides those @fgregg noted).

There should be a discussion of geographic segmentation. I'm not aware of any other non-US contributors, but the proposal doesn't take into account that user communities are largely segmented globally based on country of residence, and that this should be reflected in governance.

Less abstractly, I don't want an over-eager US committer making changes to Canada. I also don't want to have to participate in all the US approvals, etc. in order to maintain good standing. I have good processes in place for Canada, and would be responsive to Canadian PRs if there were any (I think there's only been one).

djbridges committed:
TIL how much I rely on little red squiggles

djbridges committed:
revised to incorporate first round feedback

Took another pass, adding things including: language around geographic scope/responsibilities for committers, explicit commit access revocation process, making automatic approval processes optional to adopt by country, and more.

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+``User``
+~~~~~~~~
+Anyone using OCDIDs may contribute through group discussion, reporting issues, or assisting with subject-matter expertise.
+The agreed upon communication channels for each country's community will be publicly displayed in the Open Civic Data documentation.
```
Just wanted to note that it is important for us to encourage user participation by making the adoption bar low for the technology to be used for group discussion/communication. Specifically if a user need to be on a specific platform that requires additional account creation, that may not be ideal to try to include more people who can participate on the project, that said, participating using an existing email address might be the minimum requirement.

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+--------------------
+General contributions
+~~~~~~~~~~~~~~~~~~~~~
+Any ``contributor`` may create a pull request for generative IDs, corrections, aliases, etc. 
```
This may be a high bar for many authoritative data sources (e.g. government), one of the common feedback we receive from state/county government is that they are particularly uncomfortable figuring out how github works. Will we offer additional help for creating pull requests on their behalf and/or integrate such steps as part of the contribution process?

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+``Committers`` have the ability to approve and commit pull requests from ``contributors`` and other ``committers`` within their geographic scope.
+
+* Unless otherwise specified, ``committers`` will be approved for and by their country of residence/expertise. All responsibilities and privileges associated with being a ``committer`` are limited to issues within that country's community.
+* A small subset of global ``committers`` will be established, for the sake of international coordination, whose participation may be requested in all votes, regardless of country. These global ``committers`` will not be subject to the responsibilities of communities outside their own country unless otherwise specified.
```
Does this refer specifically to the earlier use case when i18n comitters are requested for voting regardless of the country? It is not 100% clear and/or maybe I'm missing some context.

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+
+* A second ``committer`` is still required for approving government contributions, even if a government contributor becomes a ``committer`` themselves.
+* Even so, government contributors should be given wide deference within their geographic area.
+* If, because of naming conventions, geographic edge cases, etc. a government contributor requests a deviation from existing OCDID nomenclature, the community will attempt to reasonably accommodate that request (e.g. using “police_jury” as a type in lieu of “county_council” for Louisiana's county legislative body)
```
Are we okay handling these cases via aliases? That way, it may not be an entire deviation but more of a variant due to naming conventions.

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+~~~~~~~~~~~~~~~~~~~~~~~~
+Should a government entity wish to contribute to the repository, they will initially be asked to work directly with an existing ``committer`` to prepare to integrate their identifier set.
+
+* A second ``committer`` is still required for approving government contributions, even if a government contributor becomes a ``committer`` themselves.
```
In case a government contributor is committer themselves, are they still required to work directly with an existing committer to prepare the pull request (step above)? Also, it may be reasonable to require that a second committer is not within the same government org for the approval?

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+
+Government contributions
+~~~~~~~~~~~~~~~~~~~~~~~~
+Should a government entity wish to contribute to the repository, they will initially be asked to work directly with an existing ``committer`` to prepare to integrate their identifier set.
```
Is this a requirement or optional if a government contributor doesn't feel comfortable handling a PR?

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+* Even so, government contributors should be given wide deference within their geographic area.
+* If, because of naming conventions, geographic edge cases, etc. a government contributor requests a deviation from existing OCDID nomenclature, the community will attempt to reasonably accommodate that request (e.g. using “police_jury” as a type in lieu of “county_council” for Louisiana's county legislative body)
+
+Government entities (and the ``committers`` they work with) will be expected to reconcile and appropriately alias these cases with existing OCDIDs in order to ensure maximum compatibility.
```
Shall we make this a requirement? Aliasing seems to be sufficient to handle naming conventions (and potentially other use cases) for OCD-IDs that are different and/or preferable by the government EC.

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+
+Government entities (and the ``committers`` they work with) will be expected to reconcile and appropriately alias these cases with existing OCDIDs in order to ensure maximum compatibility.
+
+Identifiers created by government officials that are used in official data will be marked as such in the repository, so that developers can quickly identify the preferred identifier in case of conflict. Caution should be used, and the original submitter consulted with if possible, before changing government submitted identifiers.
```
By conflict, do we mean differences between an existing OCD-ID and a government proposed OCD-ID? Wonder if that shall have been handled by aliases per the above step. Also, by marking official data in the repository, do we have any specific mechanism/approach in mind - e.g. 'sameAsNote'?

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+Support
+--------
+``Committers`` will be expected to participate in a quarterly review of their country's new OCDIDs in order to ensure quality on-going.
+``Committers`` will be requested to contribute and maintain an ongoing style guide for creating new district types.
```
I'd make it explicit that this is more than a 'style' guide but more of a comprehensive 'how-to' guide. :)

ccongchen commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+--------
+``Committers`` will be expected to participate in a quarterly review of their country's new OCDIDs in order to ensure quality on-going.
+``Committers`` will be requested to contribute and maintain an ongoing style guide for creating new district types.
+``Committers`` will be required to participate in > 60% of their country's formal votes/actions as announced.
```
Will there be reviews to check this, wonder over which time frame this participation bar applies.

jpmckinney replied:
60% may be a high bar. I'm not sure what 'actions' refers to, but if approving PRs is an action, then it may be hard to get there, if other committers are faster than you are.

jpmckinney commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+
+A separate ``committers`` communication channel for each country-based community will be established. A list of each country's ``committer`` email addresses will be maintained and email will be used to conduct official actions unless otherwise specified. 
+
+Any individual who is a ``contributor`` and who agrees to the responsibilities listed here may request to become a ``committer``via a country's agreed upon communication channel. That request will then be shared with the group of that country's current ``committers`` for approval. 
```
Missing space after double backticks.

jpmckinney commented on proposals/drafts/ocdid_governance.rst:
```diff
@@ -0,0 +1,140 @@
+
+* There will be a two month period where existing ``committers`` may raise objections to a request, otherwise the request will be considered approved.
+* In the case that an objection is raised, a discussion and vote will be had amongst the country's existing ``committers``, with a simple majority necessary for approval.
+* A majority of a country's ``committers`` participation is necessary for a quorum on all voting matters.
```
If a country gets a large number of committers, it may be impractical to achieve a majority – and we may end up with the same situation that this OCDEP is trying to solve, which is quicker PR feedback and merging. Could we have "A majority, or #, of a country's `committers` (whichever is smaller)", where `#` is some fixed number?

jpmckinney reviewed (approved):
Looks good to me, but would want to get at least one other approval. My only concern is around the level of participation in votes/actions.

fgregg reviewed (approved):
Looks good, Donny. Thanks for your work on this.
